### DISCUSSION TITLE
Convert thrust::host_vector<float3> to std::vector<std::vector<float>>

### DISCUSSION THREAD

**Discussion by drperpen** [NVIDIA/cccl#1042] [Thrust] [answered]
Hi,
In order to use nanoflann on the result of a thrust remove_if operation on a thrust::host_vector\<float3>, I need to copy or convert it specifically to a std::vector<std::vector\<float>>. I achieved this with type float (code below) but not with float3 (code further down). Any ideas to achieve this are greatly appreciated.
Thanks!
Daniel

```
    // Generate points witn nanoflann's generateRandomPointCloud:
    std::vector<std::vector<float>> point_cloud;
    const float max_range = 20;
    generateRandomPointCloud(point_cloud, 100000, 3, max_range);

    // Create host_vector and copy the point cloud in it
    thrust::host_vector<std::vector<float>> output;
    output.resize(point_cloud.size());
    thrust::copy(output.begin(), output.end(), point_cloud.begin());

    // Send the host_vector back to std::vector<std::vector<float>>
    std::vector<std::vector<float>> backinblack;
    backinblack.resize(output.size());
    backinblack.insert(backinblack.begin(), output.begin(), output.end());
```
This below is the not working code and the error I get no matter if i use float or float3 as type in the "backinblack" std::vector:
```
    // create and populate a float3 host_vector
    thrust::host_vector<float3> float3_vector;
    for (int i = 0; i < 100; i++)
    {
        float3 xyz = make_float3(i, i + 1, i + 2);
        float3_vector.push_back(xyz);
    }

    // Send the host_vector back to std::vector<std::vector<float>>
    std::vector<std::vector<float>> backinblack;
    backinblack.resize(float3_vector.size());
    backinblack.insert(backinblack.begin(), float3_vector.begin(), float3_vector.end());
```
Error:
```
build] /usr/include/c++/9/bits/stl_algobase.h(342): error: no operator "=" matches these operands
[build]             operand types are: std::vector<float3, std::allocator<float3>> = float3
```

**Comment by miscco** [comment]
You are mixing up data types. The vector you have is a `vector<vector<float>>` but you are trying to assign a `vector<float>` to it, which should not work

**Comment by miscco** [comment, marked as answer]
I am not sure I understand the question though. Do you want to split the vector `float3_vector` into multiple smaller arrays, or do you want to just add it to `backinblack`?

If the you want the latter this should work:
```cpp
    // Send the host_vector back to std::vector<std::vector<float>>
    std::vector<std::vector<float>> backinblack{std::vector<float>{float3_vector.begin(), float3_vector.end()}};
```

This creates a vector of vector<float> via direct-list initialization of a single vector<float> that in turn is created from the range [float3_vector.begin(), float3_vector.end())

**Reply by drperpen** [reply]
Thank you so much, that is exactly what I needed!
